owners/src/owners.js

```diff
@@ -18,6 +18,129 @@ const path = require('path');
 const yaml = require('yamljs');
 
 
+/**
+ * A tree of ownership keyed by directory.
+ */
+class OwnersTree {
+  /**
+   * Constructor.
+   *
+   * @param {!string=} dirPath relative the directory containing owners file.
+   * @param {OwnersTree=} parent parent tree of the node (null for root)
+   */
+  constructor(dirPath, parent) {
+    this.parent = parent || null;
+    this.dirPath = dirPath || '.';
+
+    this.isRoot = !parent;
+
+
+    this.rules = [];
+    this.children = {};
+  }
+
+  /**
+   * Add a rule to the tree.
+   *
+   * Adds to the current node if paths match; otherwise, adds it to the node
+   * keyed by the next directory name.
+   *
+   * @param {!OwnersRule} rule rule to add.
+   */
+  addRule(rule) {
+    if (this.depth > 5) {
+      return
+    }
+    if (rule.dirPath === this.dirPath) {
+      this.rules.push(rule);
+    } else {
+      const nextDir = rule.dirPath.split(path.sep)[this.depth];
+
+      if (!this.children[nextDir]) {
+        this.children[nextDir] =
+            new OwnersTree(path.join(this.dirPath, nextDir), this);
+      }
+
+      this.children[nextDir].addRule(rule)
+    }
+  }
+
+  /**
+   * Determines the directory tree depth of the OWNERS file.
+   *
+   * Used to determine precedence. The root OWNERS file has a depth of 0.
+   *
+   * @return {number} tree depth of the OWNERS file.
+   */
+  get depth() {
+    return this.isRoot ? 0 : this.dirPath.split(path.sep).length;
+  }
+
+  /**
+   * Provides a list of ownership rules for the tree.
+   *
+   * Includes the node's own rule as well as inherited rules, in decreasing
+   * order of depth (ie. the root OWNERS rules will be last).
+   *
+   * @return {OwnersRule[]} list of all ownership rules for the directory.
+   */
+  get allRules() {
+    const parentRules = this.isRoot ? [] : this.parent.allRules;
+    return this.rules.concat(parentRules);
+  }
+
+  /**
+   * Provides a list of ownership rules applicable to a file.
+   *
+   * The most specific rules will be first, while the root-level owners rules
+   * will be last.
+   *
+   * @return {OwnersRule[]} list of rules for the file.
+   */
+  rulesForFile(filePath) {
+    let segments = filePath.split(path.sep);
+    let subtree = this;
+    
+    while (segments.length) {
+      const nextDir = segments.shift();
+      if (!subtree.children[nextDir]) {
+        break;
+      }
+      subtree = subtree.children[nextDir];
+    }
+
+    return subtree.allRules;
+  }
+
+  /**
+   * Renders the ownership tree as a string.
+   *
+   * @return {string} visual representation of the tree.
+   */
+  toString() {
+    let lines = []
+
+        const rulePrefix = '-';
+    const childPrefix = '└---';
```

Comment:
Prime real estate for emoji use ;)

Reply:
(but maybe in a followup PR, let's not test the GitHub API's handling of Unicode)

Reply:
Prime real estate for "Box-drawing character" use*
https://en.wikipedia.org/wiki/Box-drawing_character
Could have used the horizontal one instead of the dash, but I figured that would be non-obvious and harder to read/edit.